diff --git a/socialmedia.platforms.inc b/socialmedia.platforms.inc
index 3f594e8..52007fd 100644
--- a/socialmedia.platforms.inc
+++ b/socialmedia.platforms.inc
@@ -11,6 +11,7 @@ function socialmedia_socialmedia_platform_info() {
   $platforms = array_merge($platforms, socialmedia_facebook_platform_info());
   $platforms = array_merge($platforms, socialmedia_flickr_platform_info());
   $platforms = array_merge($platforms, socialmedia_googleplus_platform_info());
+  $platforms = array_merge($platforms, socialmedia_instagram_platform_info());
   $platforms = array_merge($platforms, socialmedia_linkedin_platform_info());
   $platforms = array_merge($platforms, socialmedia_pinterest_platform_info());
   $platforms = array_merge($platforms, socialmedia_rss_platform_info());
@@ -285,6 +286,73 @@ function socialmedia_googleplus_tokens($key, $profile) {
   return '';
 }
 
+function socialmedia_instagram_platform_info() {
+  $platforms = array();
+  $platforms['instagram'] = array(
+    'title' => t('Instagram'),
+    'description' => t('URL to your instagram profile or page'),
+    'homepage' => 'http://instagram.com',
+    'redirect path' => 'wtfisthis',
+    'parser callback' => 'socialmedia_instagram_parser',
+    'tokens callback' => 'socialmedia_instagram_tokens',
+  );
+  $platforms['instagram']['form'] = array(
+    'title' => t('Instagram profile'),
+    'description' => t('URL to your instagram profile'),
+  );
+  $platforms['instagram']['tokens']['multi'] = array(
+    'instagram_url' => array(
+      'name' => t("Instagram profile url"),
+      'description' => t("URL to Instagram profile."),
+    ),
+    'instagram_url-brief' => array(
+      'name' => t("Instagram profile url (brief)"),
+      'description' => t("URL to Instagram profile without protocol."),
+    ),
+    'instagram_username' => array(
+      'name' => t("Instagram name"),
+      'description' => t("Instagram profile name"),
+    ),
+  );
+  return $platforms;
+}
+
+function socialmedia_instagram_parser($values, $scope = 'site') {
+  $profile = array('url' => '', 'username' => '');
+  $str = $values['url'];
+  $pattern = '/(?:http(?:s)?:\/\/)?(www\.)?(instagram\.com\/)(\w+)+/i';
+  //$pattern = '/^http:\/\/(www\.)?twitter\.com\/(#!\/)?(?<name>[^\/]+)(/\w+)*$/';
+  if (!preg_match($pattern, $str, $matches, PREG_OFFSET_CAPTURE)) {
+    if (trim($values['url'])) {
+      form_set_error('input_instagram_url', t('Instagram URL invalid.'));
+    }
+    return FALSE;
+  }
+  $a = explode('instagram.com/', $str);
+  $profile['url'] = 'instagram.com/' . $a[1];
+  if (isset($values['username']) && trim($values['username'])) {
+    $profile['username'] = $values['username'];
+    return $profile;
+  }
+
+  form_set_error('input_instagram_username', t('Cannot derive username for this type URL. Please provide a username.'));
+  return FALSE;
+}
+
+function socialmedia_instagram_tokens($key, $profile) {
+  switch ($key) {
+  // Simple key values on the node.
+    case 'url':
+      return 'http://' . $profile['url'];
+    case 'url-brief':
+      return $profile['url'];
+    case 'username':
+      return $profile['username'];
+  }
+  return '';
+}
+
+
 function socialmedia_linkedin_platform_info() {
   $platforms = array();
   $platforms['linkedin'] = array(
@@ -754,4 +822,4 @@ function socialmedia_youtube_tokens($key, $profile) {
       return $profile['username'];
   }
   return '';
-}
\ No newline at end of file
+}
diff --git a/socialmedia.widgets.inc b/socialmedia.widgets.inc
index 359595d..c328966 100644
--- a/socialmedia.widgets.inc
+++ b/socialmedia.widgets.inc
@@ -163,6 +163,17 @@ function socialmedia_base_platforms() {
     'profile-button' => '<a href="[?profile_url=[socialmedia:sm-googleplus_url]?]" title="' . $title . '"[?link-target= target="{[socialmedia:sm-default_link_target_profile]}"?][?link-rel= rel="{[socialmedia:sm-default_link_rel_profile]}"?]><img src="[?icon-path=[socialmedia:sm-googleplus_icon-path]?]" alt="Google+ icon" ></a>',
   );
   
+  $platforms['instagram'] = array(
+    'label' => t('Instagram'),
+    'group' => t('Social media: Instagram'),
+  );
+  $title = t("Visit !varusername on Instagram", array('!varusername' => '[?profile_username=[socialmedia:sm-instagram_username]?]'));
+  $platforms['instagram']['templates'] = array(
+    'profile-url' => "[?profile_url=[socialmedia:sm-instagram_url]?]",
+    'profile-text-link' => '<a href="[?profile_url=[socialmedia:sm-instagram_url]?]" title="' . $title . '"[?link-target= target="{[socialmedia:sm-default_link_target_profile]}"?][?link-rel= rel="{[socialmedia:sm-default_link_rel_profile]}"?]>[?profile_username=[socialmedia:sm-instagram_username]?]</a>',
+    'profile-button' => '<a href="[?profile_url=[socialmedia:sm-instagram_url]?]" title="' . $title . '"[?link-target= target="{[socialmedia:sm-default_link_target_profile]}"?][?link-rel= rel="{[socialmedia:sm-default_link_rel_profile]}"?]><img src="[?icon-path=[socialmedia:sm-instagram_icon-path]?]" alt="Instagram icon" ></a>',
+  );
+
   $platforms['linkedin'] = array(
     'label' => t('LinkedIn'),
     'group' => t('Social media: LinkedIn'),
@@ -1070,6 +1081,11 @@ function socialmedia_widgets_default_sets() {
         'weight' => 1,
       ),
       array(
+        'name' => 'socialmedia_socialmedia-instagram-profile-url',
+        'data' => array(),
+        'weight' => 1,
+      ),
+      array(
         'name' => 'socialmedia_socialmedia-youtube-profile-url',
         'data' => array(),
         'weight' => 1,
@@ -1103,6 +1119,12 @@ function socialmedia_widgets_default_sets() {
         'weight' => 4,
       ),
       array(
+        'name' => 'socialmedia_socialmedia-instagram-profile-button',
+        'data' => array(),
+        'weight' => 4,
+      ),
+ 
+      array(
         'name' => 'socialmedia_socialmedia-youtube-profile-button',
         'data' => array(),
         'weight' => 5,
@@ -1178,4 +1200,4 @@ function socialmedia_widgets_default_sets() {
   );
 
   return $sets;
-}
\ No newline at end of file
+}
